frontend/coffee/global.coffee

```diff
@@ -55,42 +55,43 @@ if $(".dropdown-toggle").length > 0
         , false)
 )(box) for box in document.querySelectorAll('.upload-well')
 
+# Don't remove the follow alert from the DOM on dismiss so we can show it again
+$('#alert-follow').on 'close.bs.alert', () ->
+    $('#alert-follow').addClass 'hidden'
+    return false
+
 link.addEventListener('click', (e) ->
     e.preventDefault()
     xhr = new XMLHttpRequest()
-    follow = false
     mod_id = e.target.dataset.mod
-    if e.target.classList.contains('follow-mod-button')
-        xhr.open('POST', "/mod/#{mod_id}/follow")
-        xhr.setRequestHeader('Accept', 'application/json')
-        e.target.classList.remove('follow-mod-button')
-        e.target.classList.remove('not-following-mod')
-        e.target.classList.remove('glyphicon-star-empty')
-        e.target.classList.add('unfollow-mod-button')
-        e.target.classList.add('following-mod')
-        e.target.classList.add('glyphicon-star')
-        e.target.title = "Unfollow"
-        follow = true
-        $("#modbox-#{mod_id}-following").show()
-    else
-        xhr.open('POST', "/mod/#{mod_id}/unfollow")
-        xhr.setRequestHeader('Accept', 'application/json')
-        e.target.classList.remove('unfollow-mod-button')
-        e.target.classList.remove('following-mod')
-        e.target.classList.remove('glyphicon-star')
-        e.target.classList.add('follow-mod-button')
-        e.target.classList.add('not-following-mod')
-        e.target.classList.add('glyphicon-star-empty')
-        e.target.title = "Follow"
-        $("#modbox-#{mod_id}-following").hide()
+    xhr.open 'POST', "/mod/#{mod_id}/follow"
+    xhr.setRequestHeader 'Accept', 'application/json'
+    xhr.onload = () ->
+        $(".mod-#{mod_id}").removeClass 'not-following-mod'
+        $(".mod-#{mod_id}").addClass 'following-mod'
+        try
+            JSON.parse this.responseText
+            $('#alert-follow').removeClass 'hidden'
+        catch exc
+            window.location.href = '/register'
```

Comment:
This code is a bit weird, do you have an idea how it was supposed to work?
I guess it assumed that when anything went wrong, it would trigger an exception and not return valid JSON? And that "anything went wrong" would mean that the user doesn't have an account yet?

Since #335 at the latest the "exception->invalid JSON" conclusion is wrong. And it's likely that a problem isn't caused by the user not having an account (or even having one but not being logged in), as they wouldn't even see the follow buttons.

If we want to keep some sort of success check, we should probably check for `success == True` of the JSON. And not redirect to `/register`, but maybe display the response body somewhere?

Reply:
> This code is a bit weird, do you have an idea how it was supposed to work? I guess it assumed that when anything went wrong, it would trigger an exception and not return valid JSON? And that "anything went wrong" would mean that the user doesn't have an account yet?

Yes, I have the same guess about the intention here. I assume this line would throw at `current_user.following` without a logged in user:

https://github.com/KSP-SpaceDock/SpaceDock/blob/b7747fae7e4d9ce5f4e14481796c128060fcc2de/KerbalStuff/blueprints/mods.py#L403-L404

I'll take a look at switching to an error popup if `success` is false...

Reply:
Does it look better with the latest commit? I am still in Windows and so I am not set up for testing.

Reply:
I wonder if we should move it before toggling the classes, so they represent the probable state, and the user can basically try again by clicking the button another time?

Reply:
Good point, the toggling should only happen on success, which also then gives me an excuse to remove the `not` from the `if` line.